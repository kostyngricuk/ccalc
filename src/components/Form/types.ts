import { FieldErrors, FieldValues, GlobalError } from 'react-hook-form';

export enum EResponseStatuses {
  success,
  error
}

export interface IResponseSuccess {
  status: EResponseStatuses.success,
  message?: string
}

export interface IResponseError {
  status: EResponseStatuses.error,
  message?: string,
  errors?: FieldErrors<FieldValues> | GlobalError[]
}

export type TResponse = IResponseSuccess | IResponseError | null;