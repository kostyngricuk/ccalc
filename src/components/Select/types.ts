import { ActionMeta, GroupBase, Options, OptionsOrGroups, SingleValue } from 'react-select';


export interface ISelectOption {
  label: string;
  value: string;
}

export type TOptions = OptionsOrGroups<string, GroupBase<string>> | Options<ISelectOption> | undefined;

export type TChange = ((newValue: SingleValue<string>, actionMeta: ActionMeta<string>) => void);