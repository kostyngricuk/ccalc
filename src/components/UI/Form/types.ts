import { FieldErrors, FieldValues } from "react-hook-form";

export enum EResponseStatuses {
  success,
  error
}

export interface IResponseSuccess {
  status: EResponseStatuses.success,
  message?: string
}

export interface IResponseError {
  status: EResponseStatuses.error,
  message?: string,
  errors?: FieldErrors<FieldValues>
}

export type TResponse = IResponseSuccess | IResponseError | null;