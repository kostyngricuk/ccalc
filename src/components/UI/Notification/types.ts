enum ENotificationType {
  success,
  info
}

export default ENotificationType;