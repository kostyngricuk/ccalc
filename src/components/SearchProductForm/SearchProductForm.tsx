import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { FieldValues, useForm } from 'react-hook-form';
import { differenceBy } from "lodash";

import Title from '../UI/Title/Title';
import Button, { EnumButtonColor, EnumButtonType } from '../UI/Button/Button';
import Form from '../UI/Form/Form';
import FormField, {
  EnumFormFieldType,
} from '../UI/FormField/FormField';
import { EnumInputType, InputControlled } from '../UI/Input/Input';
import { EnumHorizontalPosition, EnumTitleVariant } from '../../services/types/global';
import { TResponse, EResponseStatuses } from '../UI/Form/types';
import { useAppDispatch, useAppSelector } from '../../services/hooks/store';
import { selectProductItems, selectProductSelectedItems } from '../../services/hooks/selectors';
import { addCustomProduct, addProduct, getProducts } from '../../services/reducers/productSlice';
import Modal from '../UI/Modal/Modal';
import { IProduct } from '../../services/types/products';
import { getKkal } from '../../services/utils/calculations';

export default function SearchProductForm() {
  const [response, setResponse] = useState<TResponse>(null);
  const [isActiveModal, setIsActiveModal] = useState<Boolean>(false);
  const { t } = useTranslation();
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch({
      type: getProducts.type,
    });
  }, []);

  const allProducts = useAppSelector(selectProductItems);
  const selectedProducts = useAppSelector(selectProductSelectedItems);

  const diffProducts = differenceBy(allProducts, selectedProducts, 'id');
  const productOptions = diffProducts.map((el) => {
    const option = {
      label: el.name,
      value: el.id
    }
    return option;
  });

  const onChangeSelectProduct = (optionValue: string) => {
    dispatch({
      type: addProduct.type,
      payload: {
        id: parseInt(optionValue, 10)
      }
    });
  }

  const {
    handleSubmit,
    control,
    reset,
    formState: { errors }
  } = useForm<FieldValues>();

  const onSubmit = handleSubmit(async (submitData: FieldValues) => {
    if (!submitData) {
      return;
    }

    const {
      name,
      proto,
      carbo,
      fats
    } = submitData;

    const product = {
      id: diffProducts.length + selectedProducts.length,
      name,
      proto,
      carbo,
      fats,
      kkal: getKkal({
        proto,
        carbo,
        fats,
      })
    } as IProduct;

    dispatch({
      type: addCustomProduct.type,
      payload: product
    });

    reset();
    setIsActiveModal(false)
  });

  useEffect(() => {
    if (Object.keys(errors).length > 0) {
      setResponse({
        status: EResponseStatuses.error,
        errors
      });
    }
  }, [errors]);

  const showModal = () => setIsActiveModal(true);
  const closeModal = () => setIsActiveModal(false);

  return (
    <>
      <FormField type={EnumFormFieldType.row}>
        <InputControlled
          type={EnumInputType.select}
          options={productOptions}
          value=""
          name="products"
          label={t('form.field.selectProduct')}
          control={control}
          onChangeTrigger={onChangeSelectProduct}
        />
        <span>{ t('calculator.form.or') }</span>
        <Button
          type={EnumButtonType.button} ariaLabel={t('calculator.form.btn.add')}
          onClick={showModal}
        >
          { t('calculator.form.btn.add') }
        </Button>
      </FormField>
        <Modal isActive={isActiveModal} onClose={closeModal}>
          <Title position={EnumHorizontalPosition.center} variant={EnumTitleVariant.h3}>{t('calculator.newProductForm.title')}</Title>
          <Form onSubmit={onSubmit} response={response}>
            <InputControlled
              required
              name="name"
              label={t('form.field.productTitle')}
              control={control}
            />
            <FormField type={EnumFormFieldType.row}>
              <InputControlled
                required
                type={EnumInputType.number}
                name="proto"
                label={t('units.proto')}
                control={control}
              />
              <InputControlled
                required
                type={EnumInputType.number}
                name="fats"
                label={t('units.fats')}
                control={control}
              />
              <InputControlled
                required
                type={EnumInputType.number}
                name="carbo"
                label={t('units.carbo')}
                control={control}
              />
            </FormField>
            <FormField type={EnumFormFieldType.actions}>
              <Button
                color={EnumButtonColor.red}
                $isOutline
                onClick={() => setIsActiveModal(false)}
              >
                {t('calculator.newProductForm.btn.cancel')}
              </Button>
              <Button type={EnumButtonType.submit}>
                {t('calculator.newProductForm.btn.submit')}
              </Button>
            </FormField>
          </Form>
        </Modal>
    </>
  )
}