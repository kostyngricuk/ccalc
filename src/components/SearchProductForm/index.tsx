import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { FieldValues, useForm } from 'react-hook-form';
import { differenceBy } from 'lodash';

import { EnumHorizontalPosition } from '@services/types/global';
import { useAppDispatch, useAppSelector } from '@services/hooks/store';
import { selectProductItems, selectProductSelectedItems } from '@services/hooks/selectors';
import { addCustomProduct, addProduct, getProducts } from '@services/reducers/productSlice';
import { IProduct, KEY_PRODUCT_ID } from '@services/types/products';
import { getKkal } from '@services/utils/calculations';
import Title from '@components/UI/Title'
import Button, { EnumButtonColor, EnumButtonType } from '@components/UI/Button'
import Form from '@components/UI/Form'
import FormField, {
  EnumFormFieldType,
} from '@components/UI/FormField'
import { InputControlled } from '@components/UI/Input'
import { TResponse, EResponseStatuses } from '@components/UI/Form/types'
import Modal from '@components/UI/Modal'
import { ISelectOption } from '@components/UI/Select/types'

export default function SearchProductForm() {
  const [response, setResponse] = useState<TResponse>(null);
  const [isActiveModal, setIsActiveModal] = useState<Boolean>(false);
  const { t } = useTranslation();
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch({
      type: getProducts.type,
    });
  }, []);

  const toggleModal = () => setIsActiveModal((prevState) => !prevState)

  const allProducts = useAppSelector(selectProductItems);
  const selectedProducts = useAppSelector(selectProductSelectedItems);

  const diffProducts = differenceBy(allProducts, selectedProducts, KEY_PRODUCT_ID);
  const productOptions: ISelectOption[] = diffProducts.map((el) => (
    {
      label: el.name,
      value: el.id.toString()
    }
  ));

  const onChangeSelectProduct = (optionValue: string) => {
    dispatch({
      type: addProduct.type,
      payload: {
        id: parseInt(optionValue, 10)
      }
    });
  }

  const {
    handleSubmit,
    control,
    reset,
    formState: { errors }
  } = useForm<FieldValues>();

  const onSubmit = handleSubmit((submitData: FieldValues) => {
    if (!submitData) {
      return;
    }

    const {
      name,
      proto,
      carbo,
      fats
    } = submitData;

    const product = {
      id: diffProducts.length + selectedProducts.length,
      name,
      proto,
      carbo,
      fats,
      kkal: getKkal({
        proto,
        carbo,
        fats,
      })
    } as IProduct;

    dispatch({
      type: addCustomProduct.type,
      payload: product
    });

    reset();
    toggleModal()
  });

  useEffect(() => {
    if (Object.keys(errors).length > 0) {
      setResponse({
        status: EResponseStatuses.error,
        errors
      });
    }
  }, [errors]);

  return (
    <>
      <FormField type={EnumFormFieldType.row}>
        <InputControlled
          type='select'
          options={productOptions}
          value=""
          name="products"
          label={t('form.field.selectProduct')}
          control={control}
          onChangeTrigger={onChangeSelectProduct}
        />
        <span>{ t('calculator.form.or') }</span>
        <Button
          type={EnumButtonType.button}
          ariaLabel={t('calculator.form.btn.add')}
          onClick={toggleModal}
        >
          { t('calculator.form.btn.add') }
        </Button>
      </FormField>
        <Modal isActive={isActiveModal} onClose={toggleModal}>
          <Title position={EnumHorizontalPosition.center} variant='h3'>{t('calculator.newProductForm.title')}</Title>
          <Form onSubmit={onSubmit} response={response}>
            <InputControlled
              required
              name="name"
              label={t('form.field.productTitle')}
              control={control}
            />
            <FormField type={EnumFormFieldType.row}>
              <InputControlled
                required
                type='number'
                name="proto"
                label={t('units.proto')}
                control={control}
              />
              <InputControlled
                required
                type='number'
                name="fats"
                label={t('units.fats')}
                control={control}
              />
              <InputControlled
                required
                type='number'
                name="carbo"
                label={t('units.carbo')}
                control={control}
              />
            </FormField>
            <FormField type={EnumFormFieldType.actions}>
              <Button
                color={EnumButtonColor.red}
                $isOutline
                onClick={() => setIsActiveModal(false)}
              >
                {t('calculator.newProductForm.btn.cancel')}
              </Button>
              <Button type={EnumButtonType.submit}>
                {t('calculator.newProductForm.btn.submit')}
              </Button>
            </FormField>
          </Form>
        </Modal>
    </>
  )
}