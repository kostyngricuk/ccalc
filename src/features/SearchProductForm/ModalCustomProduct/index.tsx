import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { FieldValues, useForm } from 'react-hook-form';

import { EnumHorizontalPosition } from 'types/global';
import { useAppDispatch, useAppSelector } from 'hooks/store';
import { selectProductSelectedItems } from 'hooks/selectors';
import { addCustomProduct } from 'store/slices/productSlice';
import { IProduct, TProducts } from 'types/products';
import { getKkal } from 'utils/calculations';
import Title from 'components/Title'
import Button, { EnumButtonColor, EnumButtonType } from 'components/Button'
import Form from 'components/Form'
import FormField, {
  EnumFormFieldType,
} from 'components/FormField'
import { InputControlled } from 'components/Input'
import { TResponse, EResponseStatuses } from 'components/Form/types'
import Modal from 'components/Modal'

export default function ModalAddCustomProduct({ diffProducts, toggleModal }: {
  diffProducts: TProducts,
  toggleModal: any
}) {
  const [response, setResponse] = useState<TResponse>(null);

  const { t } = useTranslation();
  const dispatch = useAppDispatch();

  const selectedProducts = useAppSelector(selectProductSelectedItems);

  const {
    handleSubmit,
    control,
    reset,
    formState: { errors }
  } = useForm<FieldValues>();

  const onSubmit = handleSubmit((submitData: FieldValues) => {
    if (!submitData) {
      return;
    }

    const {
      name,
      proto,
      carbo,
      fats
    } = submitData;

    const product = {
      id: diffProducts.length + selectedProducts.length,
      name,
      proto: parseInt(proto, 10),
      carbo: parseInt(carbo, 10),
      fats: parseInt(fats, 10),
      kkal: getKkal({
        proto: parseInt(proto, 10),
        carbo: parseInt(carbo, 10),
        fats: parseInt(fats, 10),
      })
    } as IProduct;

    dispatch({
      type: addCustomProduct.type,
      payload: product
    });

    reset();
    toggleModal()
  });

  useEffect(() => {
    if (Object.keys(errors).length > 0) {
      setResponse({
        status: EResponseStatuses.error,
        errors
      });
    }
  }, [errors]);


  return (
    <Modal isActive onClose={toggleModal}>
      <Title position={EnumHorizontalPosition.center} variant='h3'>{t('calculator.newProductForm.title')}</Title>
      <Form onSubmit={onSubmit} response={response}>
        <InputControlled
          required
          name="name"
          label={t('form.field.productTitle')}
          control={control}
        />
        <FormField type={EnumFormFieldType.row}>
          <InputControlled
            required
            type='number'
            name="proto"
            label={t('units.proto')}
            control={control}
          />
          <InputControlled
            required
            type='number'
            name="fats"
            label={t('units.fats')}
            control={control}
          />
          <InputControlled
            required
            type='number'
            name="carbo"
            label={t('units.carbo')}
            control={control}
          />
        </FormField>
        <FormField type={EnumFormFieldType.actions}>
          <Button
            color={EnumButtonColor.third}
            $isOutline
            onClick={toggleModal}
          >
            {t('calculator.newProductForm.btn.cancel')}
          </Button>
          <Button type={EnumButtonType.submit}>
            {t('calculator.newProductForm.btn.submit')}
          </Button>
        </FormField>
      </Form>
    </Modal>
  )
}