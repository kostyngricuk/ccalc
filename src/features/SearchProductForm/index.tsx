import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { differenceBy } from 'lodash';

import { useAppDispatch, useAppSelector } from 'hooks/store';
import { selectProductItems, selectProductSelectedItems } from 'hooks/selectors';
import { addProduct, getProducts } from 'store/slices/productSlice';
import { KEY_PRODUCT_ID } from 'types/products';
import Button, { EnumButtonType } from 'components/Button'
import FormField, {
  EnumFormFieldType,
} from 'components/FormField'
import { InputControlled } from 'components/Input'
import { ISelectOption } from 'components/Select/types'
import { FieldValues, useForm } from 'react-hook-form';
import ModalAddCustomProduct from 'features/ModalCustomProduct';

export default function SearchProductForm() {
  const [isActiveModal, setIsActiveModal] = useState<Boolean>(false);

  const { t } = useTranslation();
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch({
      type: getProducts.type,
    });
  }, []);

  const { control } = useForm<FieldValues>();

  const allProducts = useAppSelector(selectProductItems);
  const selectedProducts = useAppSelector(selectProductSelectedItems);

  const diffProducts = differenceBy(allProducts, selectedProducts, KEY_PRODUCT_ID);
  const productOptions: ISelectOption[] = diffProducts.map((el) => (
    {
      label: el.name,
      value: el.id.toString()
    }
  ));

  const onChangeSelectProduct = (optionValue: string) => {
    dispatch({
      type: addProduct.type,
      payload: {
        id: parseInt(optionValue, 10)
      }
    });
  }

  const toggleModal = () => setIsActiveModal((prevState) => !prevState)

  return (
    <>
      <FormField type={EnumFormFieldType.row}>
        <InputControlled
          type='select'
          options={productOptions}
          value=""
          name="products"
          control={control}
          label={t('form.field.selectProduct')}
          onChangeTrigger={onChangeSelectProduct}
        />
        <span>{ t('calculator.form.or') }</span>
        <Button
          type={EnumButtonType.button}
          ariaLabel={t('calculator.form.btn.add')}
          onClick={toggleModal}
        >
          { t('calculator.form.btn.add') }
        </Button>
      </FormField>
      {isActiveModal && <ModalAddCustomProduct toggleModal={toggleModal}/>}
    </>
  )
}