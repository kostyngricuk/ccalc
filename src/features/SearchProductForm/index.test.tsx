import React from "react";
import { fireEvent, render, waitFor } from "utils/test-utils";
import SearchProductForm from "features/SearchProductForm";
import { KEY_PRODUCT_ID } from "types/products";
import { defaultMockProduct, defaultMockSelectedProduct, defaultMockUser } from "constants/mocks";

describe('Search Product Form', () => {
  const state = {
    user: {
      user: defaultMockUser,
      isLoading: false,
    },
    product: {
      items: [
        defaultMockProduct,
        {
          [KEY_PRODUCT_ID]: 1,
          name: 'Bread',
          kkal: 50,
          proto: 12,
          carbo: 13,
          fats: 22,
        }
      ],
      selectedItems: [defaultMockSelectedProduct]
    }
  }
  it('Should render with default props', () => {
    const wrapper = render(<SearchProductForm />, {
      preloadedState: state
    });
    const addCustomProductBtn = wrapper.getByRole('button');
    expect(addCustomProductBtn).toBeInTheDocument();
  })

  it('Should make a spanshot', () => {
    const wrapper = render(<SearchProductForm />, {
      preloadedState: state
    });
    expect(wrapper).toMatchSnapshot();
  })

  it('Should have add custom product button', () => {
    const wrapper = render(<SearchProductForm />, {
      preloadedState: state
    });
    const addCustomProductBtn = wrapper.getByRole('button');
    expect(addCustomProductBtn).toBeInTheDocument();
  })

  it('Should work with callback on add custom product button', async () => {
    const wrapper = render(<SearchProductForm />, {
      preloadedState: state
    });
    const addCustomProductBtn = wrapper.getByRole('button');

    const mockOnClick = jest.fn();
    addCustomProductBtn.addEventListener('click', mockOnClick);
    fireEvent.click(addCustomProductBtn);
    await waitFor(() => {
      expect(mockOnClick).toHaveBeenCalledTimes(1);
    })
  })

  it('Should have a select input', () => {
    const wrapper = render(<SearchProductForm />, {
      preloadedState: state
    });
    const selectInput = wrapper.getByRole('combobox');
    expect(selectInput).toBeInTheDocument();
  })

  it('Should change value of select the product', async () => {
    const wrapper = render(<SearchProductForm />, {
      preloadedState: state
    });
    const selectInput = wrapper.getByRole('combobox');

    const mockOnChange = jest.fn();
    selectInput.addEventListener('keydown', mockOnChange);
    fireEvent.focus(selectInput);
    fireEvent.keyDown(selectInput, { key: 'ArrowDown', code: 40 });
    await waitFor(() => {
      const targetOption = wrapper.getByText('Bread');
      fireEvent.click(targetOption);
    })

    await waitFor(() => {
      expect(mockOnChange).toHaveBeenCalledTimes(1);
    });
  })

  it('Should show modal form and submit', async () => {
    const wrapper = render(<SearchProductForm />, {
      preloadedState: state
    });
    const addCustomProductBtn = wrapper.getByRole('button');

    fireEvent.click(addCustomProductBtn);
    await waitFor(() => {
      const modalForm = wrapper.getByTestId('form');
      expect(modalForm).toBeInTheDocument();
    })
  })
})