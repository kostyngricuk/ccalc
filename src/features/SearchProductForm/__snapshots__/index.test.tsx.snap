// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Product Form Should make a spanshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-eqUzNf eugqCV"
      >
        <div
          class="sc-dcJtft krnVFg Input"
        >
          <div
            class=" css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-3-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class=" css-13cymwt-control"
            >
              <div
                class=" css-1fdsijx-ValueContainer"
              >
                <div
                  class=" css-1jqq78o-placeholder"
                  id="react-select-3-placeholder"
                >
                  Select...
                </div>
                <div
                  class=" css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-3-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="react-select-3-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="products"
              type="hidden"
              value=""
            />
          </div>
        </div>
        <span>
          - OR -
        </span>
        <button
          aria-label="Add custom product"
          class="sc-aYaIB iUbBjb"
          color="primary"
          type="button"
        >
          Add custom product
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-eqUzNf eugqCV"
    >
      <div
        class="sc-dcJtft krnVFg Input"
      >
        <div
          class=" css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-3-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class=" css-13cymwt-control"
          >
            <div
              class=" css-1fdsijx-ValueContainer"
            >
              <div
                class=" css-1jqq78o-placeholder"
                id="react-select-3-placeholder"
              >
                Select...
              </div>
              <div
                class=" css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-describedby="react-select-3-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-3-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <input
            name="products"
            type="hidden"
            value=""
          />
        </div>
      </div>
      <span>
        - OR -
      </span>
      <button
        aria-label="Add custom product"
        class="sc-aYaIB iUbBjb"
        color="primary"
        type="button"
      >
        Add custom product
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "clearActions": [Function],
    "dispatch": [Function],
    "getActions": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
